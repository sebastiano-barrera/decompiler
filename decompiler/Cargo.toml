[package]
name = "decompiler"
version = "0.1.0"
edition = "2021"

[features]
trace = []
file_lock = []

[dependencies]
anyhow = "1.0.89"
bimap = "0.6.3"
gimli = { version = "0.31.1", default-features = false, features = [
    "read-all",
] }
goblin = "0.8.2"
smallvec = "1.13.2"
thiserror = "2.0.9"
rayon = "1.10.0"
slotmap = "1.0.7"
enum-assoc = "1.2.4"
arrayvec = "0.7.6"

# must be exactly 0.27.0, as following versions have MSRV 1.87.0, and we're on 1.86
facet = "=0.27.0"
facet-derive = "=0.27.0"
facet-reflect = "=0.27.0"
tracing = "0.1.41"
serde = { version = "1.0.228", features = ["derive", "rc"] }
rmp-serde = "1.3.0"

[dependencies.iced-x86]
version = "1.21.0"
default-features = false
features = ["decoder", "intel", "std", "instr_info"]

[dev-dependencies]
insta = { version = "1.40.0", default-features = false }
iced-x86 = { version = "1.21.0", features = ["encoder", "code_asm"] }
include_dir = "0.7.4"

# list ourselves here, to add the 'trace' feature when we run `cargo test` by default
decompiler = { path = ".", features = ["trace"] }
test-log = { version = "0.2.18", features = ["trace"] }

[[test]]
name = "test_tool"
path = "tests/test_tool.rs"
required-features = ["trace"]
