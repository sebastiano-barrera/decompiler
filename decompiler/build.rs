use std::process::Command;

fn main() {
    const SCRIPT_GEN_CALLCONV_TESTS: &str = "scripts/gen_callconv_tests.rb";

    println!("cargo::rerun-if-changed={}", SCRIPT_GEN_CALLCONV_TESTS);
    Command::new(SCRIPT_GEN_CALLCONV_TESTS)
        .arg("--out-c")
        .arg("test-data/x86_64_callconv.c")
        .arg("--out-rs")
        .arg("tests/generated_callconv.rs")
        .status()
        .unwrap();

    // generate the test executable
    //
    // don't use the `cc` crate because we really want to use GCC and this
    // specific set of flags
    Command::new("gcc")
        // higher levels of optimizations completely hide the calling
        // conventions
        .arg("-O1")
        .arg("-gdwarf")
        .arg("-o")
        .arg("test-data/x86_64_callconv")
        .arg("test-data/x86_64_callconv.c")
        .status()
        .unwrap();
}
