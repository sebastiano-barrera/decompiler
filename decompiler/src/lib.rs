mod ast;
mod cfg;
mod elf;
#[macro_use]
mod mil;
mod ssa;
mod tests;
mod ty;
mod x86_to_mil;

// temporarily disabled
#[cfg(any())]
mod xform;
mod xform {
    use crate::ssa;
    pub fn canonical(prog: &mut ssa::Program) {
        // do nothing!
    }

    pub fn eliminate_dead_code(prog: &mut ssa::Program) {
        let mut is_read = ssa::RegMap::for_program(prog, false);

        for bid in prog.cfg().block_ids_postorder() {
            for reg in prog.block_regs(bid).rev() {
                let mut insn = prog[reg].get();

                if insn.has_side_effects() {
                    is_read[reg] = true;
                }

                if is_read[reg] {
                    for &mut input in insn.input_regs() {
                        is_read[input] = true;
                    }
                }
            }
        }

        for (reg, &is_read) in is_read.items() {
            use crate::mil;
            if !is_read {
                prog[reg].set(mil::Insn::Void);
            }
        }
    }
}

mod util;

pub mod api;
pub mod pp;

pub use api::*;
