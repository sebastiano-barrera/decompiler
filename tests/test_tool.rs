use decompiler::{pp::PrettyPrinter, test_tool};

use include_dir::{include_dir, Dir};
use insta::assert_snapshot;
use std::{
    path::{Path, PathBuf},
    sync::OnceLock,
};

static DATA_DIR: Dir<'_> = include_dir!("$CARGO_MANIFEST_DIR/test-data/");

macro_rules! case {
    ($case:ident, $exe:expr, $func_name:ident) => {
        #[test]
        #[allow(non_snake_case)]
        fn $case() {
            let function_name = stringify!($func_name);
            let out = $exe.process_function(function_name);
            crate::assert_snapshot!(out);
        }
    };
}

macro_rules! tests_in_binary {
    ($group:ident, $exe_path:expr; $($funcs:ident),*) => {
        mod $group {
            use crate::Exe;
            static EXE: Exe = Exe::new($exe_path);

            $(case!($funcs, EXE, $funcs);)*
        }
    };
}

struct Exe {
    once_lock: OnceLock<test_tool::Result<test_tool::Tester<'static>>>,
    path: &'static str,
}
impl Exe {
    const fn new(path: &'static str) -> Self {
        Exe {
            once_lock: OnceLock::new(),
            path,
        }
    }

    fn get_or_init(&self) -> &test_tool::Tester<'static> {
        self.once_lock
            .get_or_init(|| {
                let rel_path = Path::new(self.path);
                let raw = DATA_DIR.get_file(rel_path).unwrap().contents();
                test_tool::Tester::start(raw)
            })
            .as_ref()
            .unwrap()
    }

    fn process_function(&self, function_name: &str) -> String {
        let mut buf = Vec::new();
        let mut pp = PrettyPrinter::start(&mut buf);
        self.get_or_init()
            .process_function(function_name, &mut pp)
            .expect("function decompilation failed");

        let out = String::from_utf8(buf).unwrap();
        out
    }
}

tests_in_binary!(
    exe_composite_type, "ty/test_composite_type.so";
    list_len
);
tests_in_binary!(exe_capstone, "integration/sample_capstone"; main);
tests_in_binary!(exe_matrix, "integration/sample_matrix"; sum_matrix);
tests_in_binary!(
    redis_server, "integration/redis-server";
    ctl_lookup,
    listNext,
    listTypePush
);

tests_in_binary!(
    x86_64_callconv, "x86_64_callconv";
    func000,
    func001,
    func002,
    func003,
    func004,
    func005,
    func006,
    func007,
    func008,
    func009,
    func010,
    func011,
    func012,
    func013,
    func014,
    func015,
    func016,
    func017,
    func018,
    func019,
    func020,
    func021,
    func022,
    func023,
    func024,
    func025,
    func026,
    func027,
    func028,
    func029,
    func030,
    func031,
    func032,
    func033,
    func034,
    func035,
    func036,
    func037,
    func038,
    func039,
    func040,
    func041,
    func042,
    func043,
    func044,
    func045,
    func046,
    func047,
    func048,
    func049,
    func050,
    func051,
    func052,
    func053,
    func054,
    func055,
    func056,
    func057,
    func058,
    func059,
    func060,
    func061,
    func062,
    func063,
    func064,
    func065,
    func066,
    func067,
    func068,
    func069,
    func070,
    func071,
    func072,
    func073,
    func074,
    func075,
    func076,
    func077,
    func078,
    func079,
    func080,
    func081,
    func082,
    func083,
    func084,
    func085,
    func086,
    func087,
    func088,
    func089,
    func090,
    func091,
    func092,
    func093,
    func094,
    func095,
    func096,
    func097,
    func098,
    func099,
    func100,
    func101,
    func102,
    func103,
    func104,
    func105,
    func106,
    func107,
    func108,
    func109,
    func110,
    func111,
    func112,
    func113,
    func114,
    func115,
    func116,
    func117,
    func118,
    func119,
    func120,
    func121,
    func122,
    func123,
    func124,
    func125,
    func126,
    func127,
    func128,
    func129,
    func130,
    func131,
    func132,
    func133,
    func134,
    func135,
    func136,
    func137,
    func138,
    func139,
    func140,
    func141,
    func142,
    func143,
    func144,
    func145,
    func146,
    func147,
    func148,
    func149,
    func150,
    func151,
    func152,
    func153
);
