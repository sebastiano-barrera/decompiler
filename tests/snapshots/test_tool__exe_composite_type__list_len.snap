---
source: tests/test_tool.rs
expression: out
---
list_len 0x1120+28 (file 0x1120)
            1120: 31 c0                   xor eax,eax
            1122: 48 8d 4c 24 08          lea rcx,[rsp+8]
            1127: 66 0f 1f 84 00 00 00 00 nop [rax+rax]
            1130: ff c0                   inc eax
            1132: 48 8b 49 10             mov rcx,[rcx+10h]
            1136: 48 85 c9                test rcx,rcx
            1139: 75 f5                   jne short 0000000000001130h
            113b: c3                      ret

function type: "list_len" func (p point <TypeID(17v1)>) int <TypeID(15v1)>
0 warnings.


mil program = 
program  116 instrs
    0 r12 <- Ancestral(AncestralName("stack_bottom"))
    1 r2 <- Ancestral(AncestralName("CF"))
    2 r3 <- Ancestral(AncestralName("PF"))
    3 r4 <- Ancestral(AncestralName("AF"))
    4 r5 <- Ancestral(AncestralName("ZF"))
    5 r6 <- Ancestral(AncestralName("SF"))
    6 r7 <- Ancestral(AncestralName("TF"))
    7 r8 <- Ancestral(AncestralName("IF"))
    8 r9 <- Ancestral(AncestralName("DF"))
    9 r10 <- Ancestral(AncestralName("OF"))
   10 r11 <- Ancestral(AncestralName("RBP"))
   11 r12 <- Ancestral(AncestralName("RSP"))
   12 r13 <- Ancestral(AncestralName("RIP"))
   13 r14 <- Ancestral(AncestralName("RDI"))
   14 r15 <- Ancestral(AncestralName("RSI"))
   15 r16 <- Ancestral(AncestralName("RAX"))
   16 r17 <- Ancestral(AncestralName("RBX"))
   17 r18 <- Ancestral(AncestralName("RCX"))
   18 r19 <- Ancestral(AncestralName("RDX"))
   19 r20 <- Ancestral(AncestralName("R8"))
   20 r21 <- Ancestral(AncestralName("R9"))
   21 r22 <- Ancestral(AncestralName("R10"))
   22 r23 <- Ancestral(AncestralName("R11"))
   23 r24 <- Ancestral(AncestralName("R12"))
   24 r25 <- Ancestral(AncestralName("R13"))
   25 r26 <- Ancestral(AncestralName("R14"))
   26 r27 <- Ancestral(AncestralName("R15"))
   27 r28 <- Ancestral(AncestralName("ZMM0"))
   28 r29 <- Ancestral(AncestralName("ZMM1"))
   29 r30 <- Ancestral(AncestralName("ZMM2"))
   30 r31 <- Ancestral(AncestralName("ZMM3"))
   31 r32 <- Ancestral(AncestralName("ZMM4"))
   32 r33 <- Ancestral(AncestralName("ZMM5"))
   33 r34 <- Ancestral(AncestralName("ZMM6"))
   34 r35 <- Ancestral(AncestralName("ZMM7"))
   35 r36 <- Ancestral(AncestralName("ZMM8"))
   36 r37 <- Ancestral(AncestralName("ZMM9"))
   37 r38 <- Ancestral(AncestralName("ZMM10"))
   38 r39 <- Ancestral(AncestralName("ZMM11"))
   39 r40 <- Ancestral(AncestralName("ZMM12"))
   40 r41 <- Ancestral(AncestralName("ZMM13"))
   41 r42 <- Ancestral(AncestralName("ZMM14"))
   42 r43 <- Ancestral(AncestralName("ZMM15"))
   43 r44 <- Ancestral(AncestralName("memory"))
   44 r45 <- Undefined
   45 r46 <- Undefined
   46 r47 <- Undefined
   47 r48 <- Undefined
   48 r49 <- Undefined
   49 r50 <- Undefined
   50 r51 <- Undefined
   51 r52 <- Undefined
   52 r53 <- Undefined
   53 r54 <- Undefined
   54 r55 <- Undefined
   55 r56 <- Undefined
   56 r57 <- Undefined
   57 r58 <- Undefined
   58 r59 <- Undefined
   59 r60 <- Undefined
   60 r61 <- Undefined
   61 r62 <- Undefined
   62 r63 <- Undefined
   63 r64 <- Undefined
   64 r65 <- Undefined
   65 r45 <- Ancestral(AncestralName("arg0"))
   66 r47 <- Part { src: r45, offset: 0, size: 8 }
   67 r46 <- ArithK(Add, r12, 8)
   68 r44 <- StoreMem { mem: r44, addr: r46, value: r47 }
   69 r47 <- Part { src: r45, offset: 8, size: 8 }
   70 r46 <- ArithK(Add, r12, 16)
   71 r44 <- StoreMem { mem: r44, addr: r46, value: r47 }
   72 r47 <- Part { src: r45, offset: 16, size: 8 }
   73 r46 <- ArithK(Add, r12, 24)
   74 r44 <- StoreMem { mem: r44, addr: r46, value: r47 }
0x1120:
   75 r46 <- Part { src: r16, offset: 0, size: 4 }
   76 r48 <- Part { src: r16, offset: 0, size: 4 }
   77 r46 <- Arith(BitXor, r46, r48)
   78 r49 <- Part { src: r16, offset: 4, size: 4 }
   79 r16 <- Concat { lo: r46, hi: r49 }
   80 r10 <- False
   81 r2 <- False
   82 r6 <- SignOf(r46)
   83 r5 <- IsZero(r46)
   84 r50 <- Part { src: r46, offset: 0, size: 1 }
   85 r3 <- Parity(r50)
0x1122:
   86 r45 <- Const { value: 8, size: 8 }
   87 r45 <- Arith(Add, r45, r12)
   88 r18 <- Get(r45)
0x1130:
   89 r46 <- Part { src: r16, offset: 0, size: 4 }
   90 r46 <- ArithK(Add, r46, 1)
   91 r47 <- Part { src: r16, offset: 4, size: 4 }
   92 r16 <- Concat { lo: r46, hi: r47 }
   93 r10 <- False
   94 r2 <- False
   95 r6 <- SignOf(r46)
   96 r5 <- IsZero(r46)
   97 r48 <- Part { src: r46, offset: 0, size: 1 }
   98 r3 <- Parity(r48)
0x1132:
   99 r46 <- Const { value: 16, size: 8 }
  100 r46 <- Arith(Add, r46, r18)
  101 r45 <- LoadMem { mem: r44, addr: r46, size: 8 }
  102 r18 <- Get(r45)
0x1136:
  103 r18 <- Arith(BitAnd, r18, r18)
  104 r6 <- SignOf(r18)
  105 r5 <- IsZero(r18)
  106 r47 <- Part { src: r18, offset: 0, size: 1 }
  107 r3 <- Parity(r18)
  108 r2 <- False
  109 r10 <- False
0x1139:
  110 r45 <- Not(r5)
  111 r46 <- JmpIf { cond: r45, target: 89 }
0x113b:
  112 r45 <- Void
  113 r45 <- Concat { lo: r45, hi: r16 }
  114 r45 <- Part { src: r45, offset: 0, size: 4 }
  115 r47 <- Ret(r45)


ssa pre-xform:
ssa program  122 instrs
.B0:    ;; .
  (  3)  r65 <- Ancestral(AncestralName("arg0"))
         r66 <- Part { src: r65, offset: 0, size: 8 }
  (  4)  r11 <- Ancestral(AncestralName("RSP"))
         r67 <- ArithK(Add, r11, 8)
         r43 <- Ancestral(AncestralName("memory"))
         r68 <- StoreMem { mem: r43, addr: r67, value: r66 }
         r69 <- Part { src: r65, offset: 8, size: 8 }
         r70 <- ArithK(Add, r11, 16)
         r71 <- StoreMem { mem: r68, addr: r70, value: r69 }
         r72 <- Part { src: r65, offset: 16, size: 8 }
         r73 <- ArithK(Add, r11, 24)
         r74 <- StoreMem { mem: r71, addr: r73, value: r72 }
  (  3)  r15 <- Ancestral(AncestralName("RAX"))
         r78 <- Part { src: r15, offset: 4, size: 4 }
         r76 <- Part { src: r15, offset: 0, size: 4 }
         r75 <- Part { src: r15, offset: 0, size: 4 }
         r77 <- Arith(BitXor, r75, r76)
         r79 <- Concat { lo: r77, hi: r78 }
         r118 <- Upsilon { value: r79, phi_ref: r116 }
         r86 <- Const { value: 8, size: 8 }
         r87 <- Arith(Add, r86, r11)
         r119 <- Upsilon { value: r87, phi_ref: r117 }
.B1:    ;; preds:B0,B1.
         r117 <- Phi
         r99 <- Const { value: 16, size: 8 }
         r100 <- Arith(Add, r99, r117)
  (  2)  r101 <- LoadMem { mem: r74, addr: r100, size: 8 }
  (  2)  r103 <- Arith(BitAnd, r101, r101)
         r121 <- Upsilon { value: r103, phi_ref: r117 }
  (  2)  r116 <- Phi
         r91 <- Part { src: r116, offset: 4, size: 4 }
         r89 <- Part { src: r116, offset: 0, size: 4 }
         r90 <- ArithK(Add, r89, 1)
  (  2)  r92 <- Concat { lo: r90, hi: r91 }
         r120 <- Upsilon { value: r92, phi_ref: r116 }
         r105 <- IsZero(r103)
         r110 <- Not(r105)
         r111 <- JmpIf { cond: r110, target: 89 }
.B2:    ;; preds:B1.
         r112 <- Void
         r113 <- Concat { lo: r112, hi: r92 }
         r114 <- Part { src: r113, offset: 0, size: 4 }
         r115 <- Ret(r114)


cfg:
  entry: BlockID(0)
  BlockID(0) [0:89] -> Jmp(BlockID(1))
  BlockID(1) [89:112] -> Alt { straight: BlockID(2), side: BlockID(1) }
  BlockID(2) [112:116] -> End
  domtree:
    BlockID(0)
    |  BlockID(1)
    |  |  BlockID(2)

ssa post-xform:
ssa program  122 instrs
.B0:    ;; .
  (  3)  r65 <- Ancestral(AncestralName("arg0"))
         r66 <- Part { src: r65, offset: 0, size: 8 }
  (  4)  r11 <- Ancestral(AncestralName("RSP"))
         r67 <- ArithK(Add, r11, 8)
         r43 <- Ancestral(AncestralName("memory"))
         r68 <- StoreMem { mem: r43, addr: r67, value: r66 }
         r69 <- Part { src: r65, offset: 8, size: 8 }
         r70 <- ArithK(Add, r11, 16)
         r71 <- StoreMem { mem: r68, addr: r70, value: r69 }
         r72 <- Part { src: r65, offset: 16, size: 8 }
         r73 <- ArithK(Add, r11, 24)
         r74 <- StoreMem { mem: r71, addr: r73, value: r72 }
  (  3)  r15 <- Ancestral(AncestralName("RAX"))
         r78 <- Part { src: r15, offset: 4, size: 4 }
         r76 <- Part { src: r15, offset: 0, size: 4 }
         r75 <- Part { src: r15, offset: 0, size: 4 }
         r77 <- Arith(BitXor, r75, r76)
         r79 <- Concat { lo: r77, hi: r78 }
         r118 <- Upsilon { value: r79, phi_ref: r116 }
         r87 <- ArithK(Add, r11, 8)
         r119 <- Upsilon { value: r87, phi_ref: r117 }
.B1:    ;; preds:B0,B1.
         r117 <- Phi
         r100 <- ArithK(Add, r117, 16)
  (  2)  r101 <- LoadMem { mem: r74, addr: r100, size: 8 }
         r121 <- Upsilon { value: r101, phi_ref: r117 }
  (  2)  r116 <- Phi
         r91 <- Part { src: r116, offset: 4, size: 4 }
         r89 <- Part { src: r116, offset: 0, size: 4 }
  (  2)  r90 <- ArithK(Add, r89, 1)
         r92 <- Concat { lo: r90, hi: r91 }
         r120 <- Upsilon { value: r92, phi_ref: r116 }
         r105 <- IsZero(r101)
         r110 <- Not(r105)
         r111 <- JmpIf { cond: r110, target: 89 }
.B2:    ;; preds:B1.
         r115 <- Ret(r90)



T0: {
  [pre:RSP + 8]:* := pre:arg0[8 .. 0];
  [pre:RSP + 16]:* := pre:arg0[16 .. 8];
  [pre:RSP + 24]:* := pre:arg0[24 .. 16];
  r116 := pre:RAX[8 .. 4] ++ (pre:RAX[4 .. 0] ^ pre:RAX[4 .. 0]);
  r117 := pre:RSP + 8;
  goto T1
  T1: {
    let mut r117: Bytes(8);
    let r101: Bytes(8) = [r117 + 16]:8;
    r117 := [r117 + 16]:8;
    let mut r116: Bytes(8);
    let r90: Bytes(4) = r116[4 .. 0] + 1;
    r116 := r116[8 .. 4] ++ r90;
    if ! IsZero (r101);
     {
      goto T1
    }
    Ret (r90);
    end
  }

}
